bio/bio.go

```diff
@@ -0,0 +1,323 @@
+	var records []*DataType
+	for counter := 0; counter < countN; counter++ {
+		record, err := p.Next()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				err = nil // EOF not treated as parsing error.
+			}
+			if record != nil {
+				records = append(records, record)
+			}
+			return records, err
+		}
+		records = append(records, record)
+	}
+	return records, nil
+}
+
+// Parse returns all records/reads/lines from the parser, but does not include
+// the header. It can only be called once on a given parser because it will
+// read all the input from the underlying io.Reader before exiting.
+func (p *Parser[DataType, DataTypeHeader]) Parse() ([]*DataType, error) {
+	return p.ParseN(math.MaxInt)
+}
+
+// ParseWithHeader returns all records/reads/lines, plus the header, from the
+// parser. It can only be called once on a given parser because it will read
+// all the input from the underlying io.Reader before exiting.
+func (p *Parser[DataType, DataTypeHeader]) ParseWithHeader() ([]*DataType, *DataTypeHeader, error) {
+	header, headerErr := p.Header()
+	data, err := p.Parse()
+	if headerErr != nil {
+		return data, header, err
+	}
+	if err != nil {
+		return data, header, err
+	}
+	return data, header, nil
+}
+
+/******************************************************************************
+
+Concurrent higher-level functions
+
+******************************************************************************/
+
+// ParseToChannel pipes all records/reads/lines from a parser into a channel,
+// then optionally closes that channel. If parsing a single file,
+// "keepChannelOpen" should be set to false, which will close the channel once
+// parsing is complete. If many files are being parsed to a single channel,
+// keepChannelOpen should be set to true, so that an external function will
+// close channel once all are done parsing.
+//
+// Context can be used to close the parser in the middle of parsing - for
+// example, if an error is found in another parser elsewhere and all files
+// need to close.
+func (p *Parser[DataType, DataTypeHeader]) ParseToChannel(ctx context.Context, channel chan<- DataType, keepChannelOpen bool) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			record, err := p.Next()
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					err = nil // EOF not treated as parsing error.
+				}
+				if record != nil {
+					channel <- *record
+				}
+				if !keepChannelOpen {
+					close(channel)
+				}
+				return err
+			}
+			channel <- *record
+		}
+	}
+}
+
+// ManyFastaToChannel takes a list of FASTA parsers and pipes all
+// records/reads/lines to a single channel.
+func ManyFastaToChannel(ctx context.Context, parsers []*Parser[fasta.Record, fasta.Header], channel chan<- fasta.Record) error {
+	return manyToChannel[fasta.Record, fasta.Header](ctx, parsers, channel)
+}
+
+// ManyFastqToChannel takes a list of FASTQ parsers and pipes all
```

Comment:
This functions are unnecessary as `manyToChannel` is already generic.

Reply:
You'd have to do a generic initialization on `manyToChannel` though, and I'm not sure we should have our users have to figure out generics

Reply:
Is it necessary? I think the compiler might be able to infer the types, at least my IDE was indicating that calling `manyToChannel` didn't require the generic initialization.

Reply:
Ya know what, I guess I haven't tried, since it could infer the type from the channel input. Lemme test it out, I might be WRONG and DUMB

Reply:
u right, you don't have to do generic initialization and I am just bad at generics